internal/feeder/bastion/bastion_feeder.go

```diff
@@ -0,0 +1,258 @@
+// Copyright 2024 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package bastion is an implementation of a witness feeder which talks to a bastion server.
+package bastion
+
+import (
+	"bufio"
+	"context"
+	"crypto/ed25519"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/transparency-dev/formats/log"
+	"github.com/transparency-dev/witness/internal/config"
+	"github.com/transparency-dev/witness/internal/feeder"
+	"golang.org/x/mod/sumdb/note"
+	"golang.org/x/net/http2"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/klog/v2"
+)
+
+type Config struct {
+	Addr            string
+	Prefix          string
+	Logs            []config.Log
+	WitnessKey      string
```

Comment:
This is my biggest concern about the design. Can we have the certificate passed in here instead? This would mean that we'd only have to generate the cert once, even for multiple bastion servers. It also gives us more option to generate this when the private key material isn't easily accessible to pass around in code.

Reply:
We _could_ pass it in a pre-built cert here, but I'm not sure we gain much as we still need to pass in the private key anyway in order to authenticate the witness' TLS client to the bastion host(s) (see L200 in this file).

It _might_ be possible to refactor this as taking a `crypto.Signer` rather than a direct `ed25519.PrivateKey` so that you could in theory pass in "handles", but a) I don't know whether `x509`/`tls`/`http2` make assumptions about implementations, and b) we'd still be passing around something which is equivalent to the private key.

Stepping back for a mo', I imagine the reason this is here is to authenticate the witness to the bastion at the transport layer in order to mitigate having arbitrary clients connecting and hanging to make the job of gathering cosigs harder. I guess you could try to design some sort of challenge/authentication system at the application level but that feels a bit like playing with fire when TLS can just do it for you.

Reply:
As an aside, we modified the spec to recommend using a different key for authenticating to bastions, and I've updated the code in here to reflect.